lib/spack/spack/spec.py

```diff
@@ -4143,12 +4140,13 @@ def cformat(self, *args, **kwargs):
         kwargs.setdefault('color', None)
         return self.format(*args, **kwargs)
 
-    def dep_string(self):
-        return ''.join(" ^" + dep.format() for dep in self.sorted_deps())
-
     def __str__(self):
-        ret = self.format() + self.dep_string()
-        return ret.strip()
+        spec_str = self.format()
+        dependencies = [x for x in self.traverse(root=False)]
+        dependencies.sort(key=lambda x: x.name)
+        for current_dependency in dependencies:
+            spec_str += " ^" + current_dependency.format()
+        return spec_str.strip()
```

Comment:
Is `strip()` ever needed here?  Why?

Reply:
This one was maintained from the previous implementation.

Reply:
Apparently this was needed for anonymous specs with dependencies, like `^zlib`, since otherwise they'd have a starting space.